certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,408 @@
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    uint256 key;
+    bytes32 before = at_(key);
+
+    popFront@withrevert();
+    bool success = !lastReverted;
+    
+    assert !lastReverted <=> before == at_(key - 1), "values are offset and not modified";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: popBack removes an element from the end of the queue                                          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popBack {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    bool emptyBefore = empty();
+    bytes32 backBefore = back@withrevert();
+
+    bytes32 popped = popBack@withrevert();
+    bool success = !lastReverted;
+
+    // liveness
+    assert success <=> !emptyBefore, "only fails if it's empty";
+
+    // effect
+    assert success => backBefore == popped, "previous back is returned";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: back is previous at(length - 2) after calling popBack                                                         │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popBackExpected {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+    
+    bytes32 nextBack = at_(length() - 2);
+    popBack@withrevert();
+
+    assert !lastReverted => back() == nextBack, "back is adjusted";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(x) is preserved after calling popBack                                                                     |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popBackConsistency {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    uint256 key;
+    bytes32 before = at_(key);
+
+    popBack@withrevert();
+    bool success = !lastReverted;
+    
+    assert !lastReverted <=> before == at_(key), "values are not modified";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: clear sets length to 0                                                                        │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule clear {
+    clear@withrevert();
+    bool success = !lastReverted;
+    
+    // liveness
+    assert success, "never reverts";
+    
+    // effect
+    assert success <=> length() == 0, "sets length to 0";
```

While technically correct, this is over complicated. You already `assert success`. So you can just 
```suggestion
    assert length() == 0, "sets length to 0";
```

---

I agree, thanks!